ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,218 @@
+import inspect
+from textwrap import dedent
+import collections
+import itertools
+
+import sqlalchemy as sa
+from sqlalchemy.dialects.postgresql import dialect as sa_postgres_dialect
+
+from ibis import IbisError
+import ibis.expr.rules as rlz
+from ibis.expr.signature import Argument as Arg
+from ibis.sql.postgres.compiler import (
+    PostgresUDFNode,
+    add_operation,
+    PostgreSQLExprTranslator
+)
+from ibis.sql.alchemy import _to_sqla_type
+
+
+_udf_name_cache = collections.defaultdict(itertools.count)
+
+
+class PostgresUDFError(IbisError):
+    pass
+
+
+def ibis_to_pg_sa_type(ibis_type):
+    """Map an ibis DataType to a Postgres-compatible sqlalchemy type"""
+    return _to_sqla_type(
+        ibis_type,
+        type_map=PostgreSQLExprTranslator._type_map
+    )
+
+
+def sa_type_to_postgres_str(sa_type):
+    """Map a Postgres-compatible sqlalchemy type to a Postgres-appropriate
+    string"""
+    return sa_type().compile(dialect=sa_postgres_dialect())
+
+
+def ibis_to_postgres_str(ibis_type):
+    """Map an ibis DataType to a Postgres-appropriate string"""
+    return sa_type_to_postgres_str(ibis_to_pg_sa_type(ibis_type))
+
+
+def create_udf_node(name, fields):
+    """Create a new UDF node type.
+
+    Parameters
+    ----------
+    name : str
+        Then name of the UDF node
+    fields : OrderedDict
+        Mapping of class member name to definition
+
+    Returns
+    -------
+    result : type
+        A new PostgresUDFNode subclass
+    """
+    definition = next(_udf_name_cache[name])
+    external_name = '{}_{:d}'.format(name, definition)
+    return type(external_name, (PostgresUDFNode,), fields)
+
+
+def existing_udf(name,
+                 input_types,
+                 output_type,
+                 schema=None,
+                 parameters=None):
+    """Create an ibis function that refers to an existing Postgres UDF already
+    defined in database
+
+    Parameters
+    ----------
+    name: str
+    input_types : List[DataType]
+    output_type : DataType
+    schema: str - optionally specify the schema that the UDF is defined in
+    parameters: List[str] - give names to the arguments of the UDF
+
+    Returns
+    -------
+    wrapper : Callable
```

Comment:
This can just be `Callable`. There's really no point in having a named variable in the `Returns` documentation since it's subject to change and irrelevant to users.